sdk/database/newdbplugin/middleware.go

```diff
@@ -0,0 +1,273 @@
+	}(time.Now())
+
+	mw.logger.Trace("create user",
+		"status", "started")
+	return mw.next.NewUser(ctx, req)
+}
+
+func (mw databaseTracingMiddleware) UpdateUser(ctx context.Context, req UpdateUserRequest) (resp UpdateUserResponse, err error) {
+	defer func(then time.Time) {
+		mw.logger.Trace("update user",
+			"status", "finished",
+			"err", err,
+			"took", time.Since(then))
+	}(time.Now())
+
+	mw.logger.Trace("update user", "status", "started")
+	return mw.next.UpdateUser(ctx, req)
+}
+
+func (mw databaseTracingMiddleware) DeleteUser(ctx context.Context, req DeleteUserRequest) (resp DeleteUserResponse, err error) {
+	defer func(then time.Time) {
+		mw.logger.Trace("delete user",
+			"status", "finished",
+			"err", err,
+			"took", time.Since(then))
+	}(time.Now())
+
+	mw.logger.Trace("delete user",
+		"status", "started")
+	return mw.next.DeleteUser(ctx, req)
+}
+
+func (mw databaseTracingMiddleware) Type() (string, error) {
+	return mw.next.Type()
+}
+
+func (mw databaseTracingMiddleware) Close() (err error) {
+	defer func(then time.Time) {
+		mw.logger.Trace("close",
+			"status", "finished",
+			"err", err,
+			"took", time.Since(then))
+	}(time.Now())
+
+	mw.logger.Trace("close",
+		"status", "started")
+	return mw.next.Close()
+}
+
+// ///////////////////////////////////////////////////
+// Metrics Middleware Domain
+// ///////////////////////////////////////////////////
+
+var _ Database = databaseMetricsMiddleware{}
+
+// databaseMetricsMiddleware wraps an implementation of Databases and on
+// function call logs metrics about this instance.
+type databaseMetricsMiddleware struct {
+	next Database
+
+	typeStr string
+}
+
+func (mw databaseMetricsMiddleware) Initialize(ctx context.Context, req InitializeRequest) (resp InitializeResponse, err error) {
+	defer func(now time.Time) {
+		metrics.MeasureSince([]string{"database", "Initialize"}, now)
+		metrics.MeasureSince([]string{"database", mw.typeStr, "Initialize"}, now)
+
+		if err != nil {
+			metrics.IncrCounter([]string{"database", "Initialize", "error"}, 1)
+			metrics.IncrCounter([]string{"database", mw.typeStr, "Initialize", "error"}, 1)
+		}
+	}(time.Now())
+
+	metrics.IncrCounter([]string{"database", "Initialize"}, 1)
+	metrics.IncrCounter([]string{"database", mw.typeStr, "Initialize"}, 1)
+	return mw.next.Initialize(ctx, req)
+}
+
+func (mw databaseMetricsMiddleware) NewUser(ctx context.Context, req NewUserRequest) (resp NewUserResponse, err error) {
+	defer func(start time.Time) {
+		metrics.MeasureSince([]string{"database", "CreateUser"}, start)
+		metrics.MeasureSince([]string{"database", mw.typeStr, "CreateUser"}, start)
+
+		if err != nil {
+			metrics.IncrCounter([]string{"database", "CreateUser", "error"}, 1)
+			metrics.IncrCounter([]string{"database", mw.typeStr, "CreateUser", "error"}, 1)
+		}
+	}(time.Now())
+
+	metrics.IncrCounter([]string{"database", "CreateUser"}, 1)
+	metrics.IncrCounter([]string{"database", mw.typeStr, "CreateUser"}, 1)
```

Comment:
I'm curious why these are using "CreateUser" instead of "NewUser"?

Reply:
Because I missed updating them to `NewUser`

Reply:
The alternative here is to keep it as `CreateUser` in order to keep consistency with the existing metrics.

Reply:
Ok, I wondered if we needed to keep backwards compatibility with the old metrics, but since the functions are changing I'd say it makes sense to match the metrics with the new function names.